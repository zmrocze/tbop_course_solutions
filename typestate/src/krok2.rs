use std::marker::PhantomData;

pub struct Point {
  x: i32,
  y: i32,
}

pub struct PointBuilder<S> {
  x: Option<i32>,
  y: Option<i32>,
  s : PhantomData<S>
}

pub struct XY {}
pub struct X {}
pub struct Y {}
pub struct Empty {}

impl PointBuilder<Empty> {
  pub fn new() -> Self {
    PointBuilder { x: None, y : None, s : PhantomData }
  }
  pub fn y(self, y: i32) -> PointBuilder<Y> {
    PointBuilder { x : self.x , y : Some(y), s : PhantomData }
  }
  pub fn x(self, x: i32) -> PointBuilder<X> {
    PointBuilder { x : Some(x) , y : self.y, s : PhantomData }
  }
}

impl PointBuilder<X> {
  pub fn y(self, y: i32) -> PointBuilder<XY> {
    PointBuilder { x : self.x , y : Some(y), s : PhantomData }
  }
}

impl PointBuilder<Y> {
  pub fn x(self, x: i32) -> PointBuilder<XY> {
    PointBuilder { x : Some(x) , y : self.y, s : PhantomData }
  }
}

impl PointBuilder<XY> {
  pub fn build(self) -> Point {
    Point { x : self.x.unwrap(), y : self.y.unwrap() }
  }
}

#[test]
fn works_krok2() -> Result<(), String> {
  let point2 = PointBuilder::new()
    .x(0)
    .y(1)
    .build();

  let point1 = PointBuilder::new()
    .y(1)
    .x(0)
    .build();
  println!("{} {}", point1.x, point2.y);
  Ok(())
}

#[test]
fn doesnt_compile_krok2() {
  let t = trybuild::TestCases::new();
  t.compile_fail("tests/cant_compile/unspecified_field_krok2.rs");
}