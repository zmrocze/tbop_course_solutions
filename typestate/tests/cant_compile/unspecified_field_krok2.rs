use typestate::krok2::PointBuilder;

fn doesnt_compile_krok2() {
  let point  = PointBuilder::new()
    .x(1)
    .build();
}

fn main() {}