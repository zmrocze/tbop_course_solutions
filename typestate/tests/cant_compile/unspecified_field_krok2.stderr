error[E0599]: no method named `build` found for struct `typestate::krok2::PointBuilder<X>` in the current scope
 --> tests/cant_compile/unspecified_field_krok2.rs:6:6
  |
4 |     let point  = PointBuilder::new()
  |  ________________-
5 | |     .x(1)
6 | |     .build();
  | |     -^^^^^ method not found in `PointBuilder<X>`
  | |_____|
  |
  |
  = note: the method was found for
          - `typestate::krok2::PointBuilder<XY>`
